support renamed default branches

the `ref:` input parameter seems to be hardcoded falling back to `master`.

https://github.com/actions/checkout/blob/master/README.md#L43-L46

since github support renaming the default branch, IMO the actions/checkout action should take this into account and therefore use the 'default-branch' as a fallback instead of always `master`

An example is drupal/drupal which doesn't have a master branch at all which leads to this error:
```
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +refs/heads/master:refs/remotes/origin/master
  ##[error]fatal: couldn't find remote ref refs/heads/master
  The process '/usr/bin/git' failed with exit code 128
```
So from my pint of view this is rather a bug than an enhancement.

--- a/README.md
+++ b/README.md
@@ -42,7 +42,7 @@ Refer [here](https://github.com/actions/checkout/blob/v1/README.md) for previous
 
     # The branch, tag or SHA to checkout. When checking out the repository that
     # triggered a workflow, this defaults to the reference or SHA for that event.
-    # Otherwise, defaults to `master`.
+    # Otherwise, uses the default branch.
     ref: ''
 
     # Personal access token (PAT) used to fetch the repository. The PAT is configured

--- a/__test__/input-helper.test.ts
+++ b/__test__/input-helper.test.ts
@@ -110,13 +110,6 @@ describe('input-helper tests', () => {
     )
   })
 
-  it('sets correct default ref/sha for other repo', () => {
-    inputs.repository = 'some-owner/some-other-repo'
-    const settings: IGitSourceSettings = inputHelper.getInputs()
-    expect(settings.ref).toBe('refs/heads/master')
-    expect(settings.commit).toBeFalsy()
-  })
-
   it('sets ref to empty when explicit sha', () => {
     inputs.ref = '1111111111222222222233333333334444444444'
     const settings: IGitSourceSettings = inputHelper.getInputs()

--- a/action.yml
+++ b/action.yml
@@ -8,7 +8,7 @@ inputs:
     description: >
       The branch, tag or SHA to checkout. When checking out the repository that
       triggered a workflow, this defaults to the reference or SHA for that
-      event.  Otherwise, defaults to `master`.
+      event.  Otherwise, uses the default branch.
   token:
     description: >
       Personal access token (PAT) used to fetch the repository. The PAT is configured

--- a/src/git-source-provider.ts
+++ b/src/git-source-provider.ts
@@ -19,6 +19,17 @@ export async function getSource(settings: IGitSourceSettings): Promise<void> {
   )
   const repositoryUrl = urlHelper.getFetchUrl(settings)
 
+  // Determine the default branch
+  if (!settings.ref && !settings.commit) {
+    core.startGroup('Determining the default branch')
+    settings.ref = await githubApiHelper.getDefaultBranch(
+      settings.authToken,
+      settings.repositoryOwner,
+      settings.repositoryName
+    )
+    core.endGroup()
+  }
+
   // Remove conflicting file path
   if (fsHelper.fileExistsSync(settings.repositoryPath)) {
     await io.rmRF(settings.repositoryPath)

--- a/src/github-api-helper.ts
+++ b/src/github-api-helper.ts
@@ -67,6 +67,38 @@ export async function downloadRepository(
   io.rmRF(extractPath)
 }
 
+/**
+ * Looks up the default branch name
+ */
+export async function getDefaultBranch(
+  authToken: string,
+  owner: string,
+  repo: string
+): Promise<string> {
+  return await retryHelper.execute(async () => {
+    core.info('Retrieving the default branch name')
+    const octokit = new github.GitHub(authToken)
+    const response = await octokit.repos.get({owner, repo})
+    if (response.status != 200) {
+      throw new Error(
+        `Unexpected response from GitHub API. Status: ${response.status}, Data: ${response.data}`
+      )
+    }
+
+    // Print the default branch
+    let result = response.data.default_branch
+    core.info(`Default branch '${result}'`)
+    assert.ok(result, 'default_branch cannot be empty')
+
+    // Prefix with 'refs/heads'
+    if (!result.startsWith('refs/')) {
+      result = `refs/heads/${result}`
+    }
+
+    return result
+  })
+}
+
 async function downloadArchive(
   authToken: string,
   owner: string,

--- a/src/input-helper.ts
+++ b/src/input-helper.ts
@@ -68,10 +68,6 @@ export function getInputs(): IGitSourceSettings {
         result.ref = `refs/heads/${result.ref}`
       }
     }
-
-    if (!result.ref && !result.commit) {
-      result.ref = 'refs/heads/master'
-    }
   }
   // SHA?
   else if (result.ref.match(/^[0-9a-fA-F]{40}$/)) {
@@ -110,7 +106,7 @@ export function getInputs(): IGitSourceSettings {
   core.debug(`recursive submodules = ${result.nestedSubmodules}`)
 
   // Auth token
-  result.authToken = core.getInput('token')
+  result.authToken = core.getInput('token', {required: true})
 
   // SSH
   result.sshKey = core.getInput('ssh-key')

